modules/tensor_mechanics/include/materials/czm/incremental/CZMMaterialBaseIncremental.h

```diff
@@ -0,0 +1,274 @@
+  Real _J;
+
+  /// coeffcients used to calcualte the PK1 traction increment
+  /// traction_inc = _a*((_B+_C)*_traction + D*_traction_inc)
+  ///@{
+  Real _a;
+  RankTwoTensor _B;
+  RankTwoTensor _C;
+  RankTwoTensor _D;
+  ///@}
+
+  /// coeffcients derivatives w.r.t. F. Notice that _dD_dF = _dC_dF.
+  ///@{
+  RankTwoTensor _da_dF;
+  RankFourTensor _dB_dF;
+  RankFourTensor _dC_dF;
+  ///@}
+
+  /// derivative of the rotation w.r.t F
+  RankFourTensor _dR_dF;
+
+  /// methods used to perform unsual tensor multiplications in indicail notation
+  ///@{
+  RankThreeTensor RijklVj(const RankFourTensor & R4, const RealVectorValue & V) const
+  {
+    RankThreeTensor res;
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int k = 0; k < 3; k++)
+        for (unsigned int l = 0; l < 3; l++)
+        {
+          res(i, k, l) = 0;
+          for (unsigned int j = 0; j < 3; j++)
+            res(i, k, l) += R4(i, j, k, l) * V(j);
+        }
+    return res;
+  }
+
+  RankThreeTensor RijklVl(const RankFourTensor & R4, const RealVectorValue & V) const
+  {
+    RankThreeTensor res;
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int k = 0; k < 3; k++)
+        for (unsigned int l = 0; l < 3; l++)
+        {
+          res(i, k, l) = 0;
+          for (unsigned int j = 0; j < 3; j++)
+            res(i, k, l) += R4(i, j, k, l) * V(j);
+        }
+    return res;
+  }
+
+  RankThreeTensor RijklVi(const RankFourTensor & R4, const RealVectorValue & V) const
+  {
+    RankThreeTensor res;
+    for (unsigned int j = 0; j < 3; j++)
+      for (unsigned int k = 0; k < 3; k++)
+        for (unsigned int l = 0; l < 3; l++)
+        {
+          res(j, k, l) = 0;
+          for (unsigned int i = 0; i < 3; i++)
+            res(j, k, l) += R4(i, j, k, l) * V(i);
+        }
+    return res;
+  }
+
+  RankThreeTensor RijRjkl(const RankTwoTensor & R2, const RankThreeTensor & R3) const
+  {
+    RankThreeTensor res;
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int j = 0; j < 3; j++)
+        for (unsigned int k = 0; k < 3; k++)
+          for (unsigned int l = 0; l < 3; l++)
+            res(i, k, l) += R2(i, j) * R3(j, k, l);
+    return res;
+  }
+
+  RankFourTensor RijklRjm(const RankFourTensor & R4, const RankTwoTensor & R2) const
+  {
+    RankFourTensor res;
+    res.zero();
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int k = 0; k < 3; k++)
+        for (unsigned int l = 0; l < 3; l++)
+          for (unsigned int m = 0; m < 3; m++)
+            for (unsigned int j = 0; j < 3; j++)
+              res(i, m, k, l) += R4(i, j, k, l) * R2(j, m);
+    return res;
+  }
+
+  RankThreeTensor RjkVi(const RankTwoTensor & R2, const RealVectorValue & V) const
+  {
+    RankThreeTensor res;
+
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int j = 0; j < 3; j++)
+        for (unsigned int k = 0; k < 3; k++)
+          res(i, j, k) = V(i) * R2(j, k);
+    return res;
+  }
```

Comment:
Could you move those functions to 'CohesiveZoneMethodTools"?

Reply:
I move this to `AdditionalTensorTools.h`